src/finite_field.rs

```diff
@@ -13,153 +13,218 @@ pub enum FiniteFieldError {
-        Self(SMALL_FP.mul(self.0, rhs.0))
-    }
-}
-
-impl std::ops::MulAssign for Field {
-    fn mul_assign(&mut self, rhs: Self) {
-        *self = *self * rhs;
-    }
-}
-
-impl std::ops::Div for Field {
-    type Output = Field;
-
-    fn div(self, rhs: Self) -> Self {
-        self * rhs.inv()
-    }
-}
-
-impl std::ops::DivAssign for Field {
-    fn div_assign(&mut self, rhs: Self) {
-        *self = *self / rhs;
-    }
-}
-
-impl PartialEq for Field {
-    fn eq(&self, rhs: &Self) -> bool {
-        SMALL_FP.from_elem(self.0) == SMALL_FP.from_elem(rhs.0)
-    }
-}
-
-impl Eq for Field {}
-
-impl Field {
-    /// Modular exponentation
-    pub fn pow(self, exp: Self) -> Self {
-        Self(SMALL_FP.pow(self.0, SMALL_FP.from_elem(exp.0)))
-    }
-
-    /// Modular inverse
-    ///
-    /// Note: inverse of 0 is defined as 0.
-    pub fn inv(self) -> Self {
-        Self(SMALL_FP.inv(self.0))
-    }
-}
-
-impl From<u32> for Field {
-    fn from(x: u32) -> Self {
-        Field(SMALL_FP.elem(x as u128))
-    }
-}
-
-impl From<Field> for u32 {
-    fn from(x: Field) -> Self {
-        SMALL_FP.from_elem(x.0) as u32
-    }
-}
-
-impl PartialEq<u32> for Field {
-    fn eq(&self, rhs: &u32) -> bool {
-        SMALL_FP.from_elem(self.0) == *rhs as u128
-    }
+/// Objects with this trait represent an element of `GF(p)` for some prime `p`.
+pub trait FieldElement:
+    Sized
+    + PartialEq
+    + Eq
+    + std::ops::Add
+    + std::ops::AddAssign
+    + std::ops::Sub
+    + std::ops::SubAssign
+    + std::ops::Mul
+    + std::ops::MulAssign
+    + std::ops::Div
+    + std::ops::DivAssign
+    + std::ops::Neg
+    + std::fmt::Display
+{
+    /// The integer representation of the field element.
+    type Integer;
+
+    /// Modular exponentation, i.e., `self^exp (mod p)`.
+    fn pow(self, exp: Self) -> Self;
+
+    /// Modular inversion, i.e., `self^-1 (mod p)`. If `self` is 0, then the output is undefined.
+    fn inv(self) -> Self;
+
+    /// Returns the prime modulus `p`.
+    fn modulus() -> Self::Integer;
+
+    /// Returns a generator of the multiplicative subgroup of size `FieldElement::num_roots()`.
+    fn generator() -> Self;
+
+    /// Returns the size of the multiplicative subgroup generated by `FieldElement::generator()`.
+    fn num_roots() -> Self::Integer;
 }
 
-impl std::fmt::Display for Field {
-    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
-        write!(f, "{}", SMALL_FP.from_elem(self.0))
-    }
+macro_rules! make_field {
+    (
+        $(#[$meta:meta])*
+        $elem:ident, $int:ident, $fp:ident
+    ) => {
+        $(#[$meta])*
+        #[derive(Clone, Copy, Debug, PartialOrd, Ord, Hash, Default)]
+        pub struct $elem(u128);
+
+        impl PartialEq for $elem {
+            fn eq(&self, rhs: &Self) -> bool {
+                $fp.from_elem(self.0) == $fp.from_elem(rhs.0)
+            }
+        }
+
+        impl Eq for $elem {}
+
+        impl std::ops::Add for $elem {
```

Comment:
1. I like this idea, but I would prefer something more like this:
```rust
         impl Add for &$elem {
             type Output = $elem;
             fn add(self, rhs: Self) -> $elem {
                 *self + *rhs
             }
         }
```
I like this better because it re-uses the implementation of `Add` for `$elem`.

2. What's the analogous patter for the assignment operators, e.g., `AddAssign for &$elem`? Does this even make sense?

Reply:
1. I agree, yours is better code reuse.
2. `std::ops::AddAssign::add_assign` explicitly takes `&mut self` and returns nothing (which makes sense, the point of `+=` is to mutate the value) so I don't think it makes sense to replicate the pattern there. I think the extra `impl`s are only necessary/helpful for those traits in `std::ops` where the method receiver is a value.

Reply:
Done, for Add, Sub, Mul, and Div, and Neg.